keps/sig-scheduling/20190926-default-even-pod-spreading.md

```diff
@@ -0,0 +1,278 @@
+
+On the other hand, cluster operators know the underlying topology of the cluster, which makes
+them suitable to provide default spreading rules for all workloads in their cluster.
+
+### Goals
+
+- Cluster operators can define default spreading rules for pods that don't provide any.
+- Workloads are spread with the default rules if they belong to the same service, controller,
+replica set or stateful set,
+and if they don't define `TopologySpreadConstraints`.
+
+### Non-Goals
+
+- Removal of `SelectorSpreadPriority`, `ServiceSpreadingPriority` or `ServiceAntiAffinity` priorities.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to set default spreading rules for workloads in the cluster.
+Currently, `SelectorSpreadPriority` provides a canned priority that spreads across nodes
+and zones (`failure-domain.beta.kubernetes.io/zone`). However, the nodes in my cluster have
+custom topology keys (for physical host, rack, etc.).
+
+#### Story 2
+
+As a workload author, I want to spread the workload in the cluster, but:
+(1) I don't know the topology of the cluster I'm running on.
+(2) I want to be able to run my PodSpec in different clusters (on-prem and cloud).
+
+### Implementation Details/Notes/Constraints
+
+Note that a priority given by default `EvenPodsSpreading` rules could conflict with
+`SelectorSpreadingPriority`.
+Operators can disable `SelectorSpreadingPriority`. But once default rules for `EvenPodsSpreading` is GA,
+we can consider removing `SelectorSpreadingPriority` and replacing it by an equivalent
+k8s default to the default rules for `EvenPodsSpreading`.
+
+### Risks and Mitigations
+
+`EvenPodsSpreading` has some overhead and we currently ensure that pods that don't use the
+feature get minimally affected. After default rules for `EvenPodsSpreading` is rolled out,
+all pods will run through the algorithms.
+However, we should ensure that the running overhead is not significantly higher than
+`SelectorSpreadingPriority` if using the k8s default.
+
+## Design Details
+
+### API
+
+A new structure called `TopologySpreadConstraint` is introduced to `KubeSchedulerConfiguration`.
+
+```go
+type KubeSchedulerConfiguration struct {
+	....
+	// DefaultTopologySpreadConstraints defines spreading constraints to be applied to pods
+	// that don't define any.
+	// If not specified, the scheduler applies the following default.
+	// +optional
+	DefaultTopologySpreadConstraints []TopologySpreadConstraint
+	....
+}
+
+// TopologySpreadConstraint specifies how to spread pods among the given topology.
+// Pod selectors are deduced from the resource definitions that the pod belongs to
+// (includes services, controllers, replica sets and stateful sets).
+type TopologySpreadConstraint struct {
+	MaxSkew int32
+	TopologyKey string
+	WhenUnsatisfiable corev1.UnsatisfiableConstraintAction
+}
+```
+
+Note that `TopologySpreadConstraint` is similar to `k8s.io/api/core/v1.TopologySpreadConstraint`,
```

Comment:
As current `TopologySpreadConstraint.LabelSelector` is optional, you can reuse that structure. (validation logic can help detect illegal presence of labelSelector)

If you still want to create a new structure, the name should be different from `TopologySpreadConstraint` to avoid confusion.

Reply:
> As current `TopologySpreadConstraint.LabelSelector` is optional, you can reuse that structure. (validation logic can help detect illegal presence of labelSelector)

@Huang-Wei  What happens currently in EvenPodsSpreading when the label selector is not specified?

Reply:
These are different APIs, so I think they can be named the same. But I'm happy to reuse the struct if the selector is indeed optional. And during validation we ensure that it was not set.

Reply:
@ahg-g If `LableSelector` is nil, no pods will be matched, then the incoming pod can land on any node which has the topologyKey(s) present. So in other words, it doesn't fit this KEP's requirement. But we _can_ modify its semantics and document it.